contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,188 @@
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(MINTER_ADDRESS);
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable sourceContractAddress;
+    // We will not mint the first `tokenReserve` tokens sent to this subnet.
+    // This should be constructed to match the initial token supply of this subnet.
+    // This will mean we will not mint tokens until the source contact is collateralized.
+    uint256 public tokenReserve;
+
+    // Used for sending an receiving Teleporter messages.
+    ITeleporterMessenger public immutable teleporterMessenger;
+
+    // TODO we probably want to add the original token supply from this chain to the constructor.
+    constructor(
+        address teleporterMessengerAddress,
+        bytes32 sourceBlockchainID_,
+        address sourceContractAddress_,
+        uint256 tokenReserve_
+    ) {
+        currentBlockchainID = WarpMessenger(WARP_PRECOMPILE_ADDRESS)
+            .getBlockchainID();
+
+        require(
+            teleporterMessengerAddress != address(0),
+            "Invalid Teleporter Messenger Address"
+        );
+        teleporterMessenger = ITeleporterMessenger(teleporterMessengerAddress);
+
+        require(sourceBlockchainID_ != bytes32(0), "Invalid Source Chain ID");
+        require(
+            sourceBlockchainID_ != currentBlockchainID,
+            "Cannot Bridge With Same Blockchain"
+        );
+        sourceBlockchainID = sourceBlockchainID_;
+
+        require(
+            sourceContractAddress_ != address(0),
+            "Invalid Source Contract Address"
+        );
+        sourceContractAddress = sourceContractAddress_;
+
+        tokenReserve = tokenReserve_;
+    }
+
+    /**
+     * @dev See {ITeleporterReceiver-receiveTeleporterMessage}.
+     *
+     * Receives a Teleporter message and routes to the appropriate internal function call.
+     */
+    function receiveTeleporterMessage(
+        bytes32 senderBlockchainID,
+        address senderAddress,
+        bytes calldata message
+    ) external nonReentrant {
+        // Only allow the Teleporter messenger to deliver messages.
+        require(
+            msg.sender == address(teleporterMessenger),
+            "Unauthorized teleporter contract"
+        );
+
+        // Only allow messages from the source chain.
+        require(
+            senderBlockchainID == sourceBlockchainID,
+            "Invalid Source Chain"
+        );
+
+        // Only allow the partner contract to send messages.
+        require(senderAddress == sourceContractAddress, "Unauthorized Sender");
+
+        (address recipient, uint256 amount) = abi.decode(
```

Comment:
is amount 0 valid? Right now we'll emit `CollateralAdded` or amount 0, and when both token reserve and amount are 0 we'll emit `MintNativeTokens`

Reply:
Added a check here.